<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Yang Shuai" />

<meta name="date" content="2015-12-23" />

<title>Bionexr: a R package for integrative network-based analysis of cancer somatic mutation and expression data</title>




<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0A%7D%0Apre%20%7B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>



<div id="header">
<h1 class="title">Bionexr: a R package for integrative network-based analysis of cancer somatic mutation and expression data</h1>
<h4 class="author"><em>Yang Shuai</em></h4>
<h4 class="date"><em>2015-12-23</em></h4>
</div>


<p>Cancer genome projects have generated massive genome and transcriptome sequencing data, which makes tumor-specific alterations such as somatic mutation and gene expression information easily available. Due to the lack of tools for subsequently integrating these complex information, we implement an R package “bionexr” that uses network to integrate and analyze cancer somatic mutation and expression data. Bionexr provides these features:</p>
<ul>
<li>A protein-protein interaction (PPI)-based approach</li>
<li>A pathway-based appoach</li>
<li>Visualization of the results</li>
</ul>
<div id="quick-start" class="section level2">
<h2>Quick Start</h2>
<ul>
<li>First, run the following instructions. Note that if it is the first time to use bionexr, <code>prepare_ma()</code> will take some time to download the dataset used by gene module.</li>
</ul>
<pre><code>library(bionexr)
prepare_ma() # prepare data used by bionexr</code></pre>
<ul>
<li>For PPI-based approach, follow the instructions below:</li>
</ul>
<pre><code>res.gene &lt;- perform_gene_ppi(hnsc_mut_part, hnsc_exp_part)
res.network &lt;- perform_network_ppi(res.gene[[2]], res.gene[[3]])
g &lt;- network_from_ppi(res.network)
plot_ppi(g)</code></pre>
<p>And the result would look like below:</p>
<ul>
<li>For pathway-based approach, follow the instructions below:</li>
</ul>
<pre><code>res.gene &lt;- perform_gene_pathway(hnsc_mut_part, hnsc_exp_part)
res.network &lt;- perform_network_pathway(res.gene[[2]], res.gene[[3]], hnsc_expressed_genes)

g &lt;- network_from_significant_branches(res.network)
plot_pathway(g)</code></pre>
<p>And the result would look like below:</p>
</div>
<div id="usage" class="section level2">
<h2>Usage</h2>
<div id="download-cancer-genome-data-using-data-download-module" class="section level3">
<h3>Download cancer genome data using “Data Download” module</h3>
<p><code>firehose_get</code> is the main command to download cancer genome data from firehose. Here we use <code>firehose_get</code> to download HNSC data.</p>
<pre><code>mut_data &lt;- firehose_get(&quot;HNSC&quot;, &quot;mutation&quot;, run_date = &quot;2015_08_21&quot;, run_type = &quot;stddata&quot;)
mut_data &lt;- mut_data[[1]]
mut_sample_ids &lt;- unique(mut_data[[7]])

exp_data &lt;- firehose_get(&quot;HNSC&quot;, &quot;expression&quot;, run_date = &quot;2015_08_21&quot;, run_type = &quot;stddata&quot;)
exp_data &lt;- exp_data[[1]]
exp_sample_ids &lt;- colnames(exp_data)

common_case &lt;- intersect(mut_sample_ids, exp_sample_ids)
exp_control &lt;- grepl(&quot;-11$&quot;, exp_sample_ids)
hnsc_mut &lt;- mut_data[mut_data[[7]] %in% common_case, ]
hnsc_exp &lt;- exp_data[, (exp_sample_ids %in% common_case) | exp_control]</code></pre>
</div>
<div id="using-gene-analysis-module" class="section level3">
<h3>Using “Gene Analysis” module</h3>
<p><code>perform_gene_ppi</code> and <code>perform_gene_pathway</code> are the two main commands for performing “Gene Analysis”. As you can guess from the function name, <code>perform_gene_ppi</code> is for PPI-based approach and <code>perform_gene_pathway</code> is for pathway-based approach.</p>
<p>See the instructions below, note that <code>hnsc_mut</code> and <code>hnsc_exp</code> are from “Data Download” module:</p>
<pre><code># For PPI-based approach
ppi.gene &lt;- perform_gene_ppi(hnsc_mut, hnsc_exp)
# For pathway-based approach
pathway.gene &lt;- perform_gene_pathway(hnsc_mut, hnsc_exp)</code></pre>
<p>Note that before performing “Gene Analysis”, run command <code>prepare_ma()</code> first. This module would take a few time to finish, drink some coffee happily. ### Using “Network Analysis” module <code>perform_network_ppi</code> and <code>perform_network_pathway</code> are the two main commands for performing “Network Analysis”. As the same to “Gene Analysis” module, <code>perform_network_ppi</code> is for PPI-based approach and <code>perform_network_pathway</code> is for pathway-based approach.</p>
<p>See the instructions below, note that <code>hnsc_exp</code> is from “Data Download” module, and <code>ppi.gene</code> and <code>pathway.gene</code> are from “Gene Analysis” module:</p>
<pre><code># For PPI-based approach
ppi.network &lt;- perform_network_ppi(ppi.gene[[2]], ppi.gene[[3]])
# For pathway-based approach
expressed_genes &lt;- identify_expressed_genes(hnsc_exp)
pathway.network &lt;- perform_network_pathway(pathway.gene[[2]], pathway.gene[[3]], expressed_genes)</code></pre>
</div>
<div id="using-visualization-module" class="section level3">
<h3>Using “Visualization” module</h3>
<p><code>plot_ppi</code> and <code>plot_pathway</code> are the two main commands for performing “Visualization” module.<code>plot_ppi</code> is for PPI-based approach’s result and <code>plot_pathway</code> is for pathway-based approach’s result.</p>
<p>See the instructions below, note that <code>ppi.network</code> and <code>pathway.network</code> are from “Network Analysis” module:</p>
<pre><code># For PPI-based approach's result
ppi.g &lt;- network_from_ppi(ppi.network)
plot_ppi(ppi.g)
# For pathway-based approach's result
pathway.g &lt;- network_from_significant_branches(pathway.network)
plot_pathway(pathway.g)</code></pre>
</div>
<div id="using-gene-analysis-and-network-analysis-in-one-command" class="section level3">
<h3>Using “Gene Analysis” and “Network Analysis” in one command</h3>
<p>The commands <code>perform_main_ppi</code> and <code>perform_main_pathway</code> can perform “Gene Analysis” and “Network Analysis”, and the result can be visualized straightforward.</p>
<p>The example instructions are written below, note that <code>hnsc_mut</code> and <code>hnsc_exp</code> are from “Data Download” module:</p>
<pre><code># For PPI-based approach
prepare_ma()
ppi.res &lt;- perform_main_ppi(hnsc_mut, hnsc_exp, jobname = &quot;HNSC&quot;, use_cache = TRUE)
# Visualize PPI-based result
ppi.g &lt;- network_from_ppi(ppi.res)
plot_ppi(ppi.g)

# For pathway-based approach
prepare_ma()
pathway.res &lt;- perform_main_pathway(hnsc_mut, hnsc_exp, jobname = &quot;HNSC&quot;, test = TRUE)
pathway.g &lt;- network_from_significant_branches(pathway.res)
plot_pathway(pathway.g)</code></pre>
</div>
</div>
<div id="contact" class="section level2">
<h2>Contact</h2>
<p>Please send email to <a href="mailto:yannis.pku@gmail.com">yannis.pku@gmail.com</a> if you have any questions.</p>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
